File: python/leetcodehard/median_of_two_sorted_arrays.py
"""
https://leetcode.com/problems/median-of-two-sorted-arrays/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def findMedianSortedArrays(self, nums1, nums2):
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        m, n = len(nums1), len(nums2)
        low, high = 0, m
        half = (m + n + 1) // 2

        while low <= high:
            i = (low + high) // 2
            j = half - i

            left1 = nums1[i - 1] if i > 0 else float("-inf")
            right1 = nums1[i] if i < m else float("inf")
            left2 = nums2[j - 1] if j > 0 else float("-inf")
            right2 = nums2[j] if j < n else float("inf")

            if left1 <= right2 and left2 <= right1:
                if (m + n) % 2 == 0:
                    return (max(left1, left2) + min(right1, right2)) / 2
                return max(left1, left2)
            elif left1 > right2:
                high = i - 1
            else:
                low = i + 1

        raise ValueError("Input arrays are not sorted")


if __name__ == "__main__":
    test_input_1 = ([1, 3], [2])
    output = Solution().findMedianSortedArrays(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ([1, 2], [3, 4])
    output = Solution().findMedianSortedArrays(*test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/medianOfTwoSortedArrays.ts
/*
https://leetcode.com/problems/median-of-two-sorted-arrays/

status - completed
*/

import { logp } from "../utils/logger";

function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
  let a = nums1;
  let b = nums2;
  if (a.length > b.length) {
    a = nums2;
    b = nums1;
  }

  const m = a.length;
  const n = b.length;
  const half = Math.floor((m + n + 1) / 2);
  let low = 0;
  let high = m;

  while (low <= high) {
    const cutA = Math.floor((low + high) / 2);
    const cutB = half - cutA;

    const leftA = cutA === 0 ? -Infinity : a[cutA - 1];
    const leftB = cutB === 0 ? -Infinity : b[cutB - 1];
    const rightA = cutA === m ? Infinity : a[cutA];
    const rightB = cutB === n ? Infinity : b[cutB];

    if (leftA <= rightB && leftB <= rightA) {
      if ((m + n) % 2 === 0) {
        return (Math.max(leftA, leftB) + Math.min(rightA, rightB)) / 2;
      }
      return Math.max(leftA, leftB);
    } else if (leftA > rightB) {
      high = cutA - 1;
    } else {
      low = cutA + 1;
    }
  }

  throw new Error("input arrays are not sorted");
}

logp(`input [1,3] [2] expected 2 actual ${findMedianSortedArrays([1, 3], [2])}`);
logp(`input [1,2] [3,4] expected 2.5 actual ${findMedianSortedArrays([1, 2], [3, 4])}`);
logp(`input [] [1] expected 1 actual ${findMedianSortedArrays([], [1])}`);
